### ISSUE TITLE
`xarray` data objects not being accounted in the worker's managed memory

### ISSUE THREAD

**Issue by mahendrapaipuri** [dask/distributed#5383]
**What happened**: When we use `xarray` data objects in the dask computations, they are not being accounted in the worker's managed memory. We see only the size of reference tables of `xarray` is being shown in the managed memory.

**What you expected to happen**: As `xarray` is one of the standard library, we would like to see these data objects being accounted in the managed memory. As workers can only spill the managed memory to disk, we faced number of issues when running out of memory in our workflows.

**Minimal Complete Verifiable Example**:

```
import time
import xarray as xr

from dask import delayed
from distributed import Client, wait
from dask.sizeof import sizeof

# Initialise sample data
def init_data(n):
    return xr.Dataset( {"testdata_1": (["a", "b", "y", "x"],
    np.random.rand(1, 1, n, n))} , {"testdata_2": (["a", "b", "y", "x"],
    np.random.rand(1, 1, n, n))} )

# Get size of objects using dask sizeof
def get_size():
    data_obj = init_data(1024)
    obj_size = sizeof(data_obj)
    print("xarray data object size is {} MiB".format(obj_size / (1024 * 1024)))

# Sample operations
def sample_operations(ds):
    time.sleep(30)
    ds = ds * 2 + 1.5
    return ds

if __name__ == "__main__":
    get_size()
    print("Creating scheduler and 4 workers")
    client = Client(n_workers=4, threads_per_worker=1)
    data = delayed(init_data)(1024)
    graph = delayed(sample_operations)(data)
    future = client.compute(graph, sync=True)
    wait(future)
```

**Anything else we need to know?**: The origin of this issue is in the dask's `sizeof()` implementation, `xarray` is not included. If we register `xarray` in the dispatcher function, we will fix this issue.
```
@sizeof.register_lazy("xarray")
def register_xarray():
    import xarray

    @sizeof.register(xarray.core.dataset.Dataset)
    def sizeof_xarray_dataset(x):
        return int(x.nbytes)
```

**Environment**:

- Dask version: 2021.9.0
- Python version: 3.8.10
- Operating System: Debian GNU/Linux 10
- Install method (conda, pip, source): pip

**Comment by martindurant** [comment]
There is an additional wrinkle here: xarray object are generally lazy and don't include their arrays in memory until they are explicitly loaded. Indeed, those can be dask-array (although this is less likely as a result of a dask task, but still possible). So we need to know how much of those `nbytes` are in memory. I trust this is possible, but I don't know how.

**Comment by jrbourbeau** [comment]
Thanks for reporting @mahendrapaipuri. I agree it would be good for Dask to better understand the size of `xarray` objects. There's a discussion downstream in `xarray` (xref https://github.com/pydata/xarray/issues/5764) about adding support for this directly in `xarray`

**Comment by jakirkham** [comment]
It's worth noting that the XArray issue was closed out as not implemented. Here is the rationale ( https://github.com/pydata/xarray/issues/5764#issuecomment-933822350 )

To Stephan's point (please see linked comment), maybe it is worth considering a Dask specific protocol (like `__dask_sizeof__`) and using that

**Comment by fjetter** [comment]
I'm mildly reluctant adding another dunder method for this considering that we have the `sizeof` dispatch that is used for everything else. I would expect this to work for xarray as well

**Comment by mrocklin** [comment]
To be clear, what @jakirkham is suggesting is possible, there is a dask-specific protocol, it's the `dask.sizeof` function.  https://docs.dask.org/en/stable/how-to/extend-sizeof.html

If Xarray developers wanted to implement that that could work well.

**Comment by jakirkham** [comment]
Good point Matt. Wasn't thinking about that option. Agree that's a good choice
